resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java

```diff
@@ -1,32 +1,140 @@
+      OutputStream stream;
+      public AsyncOperation(final OutputStream stream)
+      {
+         this.stream = stream;
+      }
+      public void work(ServletOutputStream sos) {
+         ResteasyContext.pushContextDataMap(contextDataMap);
+         try {
+            doWork(sos);
+         }finally {
+            ResteasyContext.removeContextDataLevel();
+         }
+      }
+      protected abstract void doWork(ServletOutputStream sos);
+   }
+
+   public class WriteOperation extends AsyncOperation
+   {
+
+      private byte[] bytes;
+      private int offset;
+      private int length;
+
+      public WriteOperation(final OutputStream stream, final byte[] bytes, final int offset, final int length)
+      {
+         super(stream);
+         this.bytes = bytes;
+         this.offset = offset;
+         this.length = length;
+      }
+
+      @Override
+      protected void doWork(ServletOutputStream sos)
+      {
+         try
+         {
+            stream.write(bytes, offset, length);
+            // we only are complete if isReady says we're good to write, otherwise
+            // we will be complete in the next onWritePossible or onError
+            if(sos == null || sos.isReady()) {
+               future.complete(null);
+            }
+         } catch (IOException e)
+         {
+            future.completeExceptionally(e);
+         }
+      }
+
+      @Override
+      public String toString()
+      {
+         return "[write: "+new String(bytes)+"]";
+      }
+   }
+
+   public class FlushOperation extends AsyncOperation
+   {
+
+      public FlushOperation(final OutputStream os)
+      {
+         super(os);
+      }
+
+      @Override
+      protected void doWork(ServletOutputStream sos)
+      {
+         try
+         {
+            stream.flush();
+            // we only are complete if isReady says we're good to write, otherwise
+            // we will be complete in the next onWritePossible or onError
+            if(sos == null || sos.isReady()) {
+               future.complete(null);
+            }
+         } catch (IOException e)
+         {
+            future.completeExceptionally(e);
+         }
+      }
+
+      @Override
+      public String toString()
+      {
+         return "[flush]";
+      }
+   }
+
    protected HttpServletResponse response;
    protected int status = 200;
    protected MultivaluedMap<String, Object> outputHeaders;
    protected ResteasyProviderFactory factory;
    protected OutputStream outputStream = new DeferredOutputStream();
+   protected HttpServletRequest request;
+   protected Map<Class<?>, Object> contextDataMap;
 
    /**
     * RESTEASY-684 wants to defer access to outputstream until a write happens
     *
     */
-   protected class DeferredOutputStream extends OutputStream
+   protected class DeferredOutputStream extends AsyncOutputStream implements WriteListener
    {
+      private boolean asyncRegistered;
+      private Queue<AsyncOperation> asyncQueue;
```

Comment:
I know this is an old-ish PR, but we are working on a possible bug in this area.  I'm trying to understand why a queue is needed.  Aren't clients only supposed to call `asyncWrite` when a previous `asyncWrite` or `asyncFlush` has completed?  I realize there is no guarantee on what callers will do, and the queue could possibly handle weird cases.  Is that why a queue was used?  Trying to understand if it's needed if the caller is chaining them with the `CompletionStage` api (like [here](https://github.com/resteasy/resteasy/blob/6e94de442cd2aac93207c4e1fc06a1bbb214c1d0/resteasy-core/src/main/java/org/jboss/resteasy/plugins/providers/ProviderHelper.java#L164)).

Anyhow, I know this code has a changed a bit.  If this is better asked in some other way, just let me know.